core/src/main/java/org/infinispan/interceptors/distribution/TriangleDistributionInterceptor.java

```diff
@@ -324,37 +356,36 @@ private Object primaryOwnerWrite(InvocationContext context, DataWriteCommand com
          final CommandInvocationId id = dwCommand.getCommandInvocationId();
-         if (!dwCommand.isSuccessful()) {
-            if (trace) {
-               log.tracef("Command %s not successful in primary owner.", id);
-            }
+         if (!logAndCheckSuccessful(dwCommand)) {
             return;
          }
-         if (distributionInfo.owners().size() > 1) {
-            Collection<Address> backupOwners = distributionInfo.backups();
-            if (rCtx.isOriginLocal() && (isSynchronous(dwCommand) || dwCommand.isReturnValueExpected())) {
-               commandAckCollector.create(id, rv, distributionInfo.owners(), dwCommand.getTopologyId());
-               //check the topology after registering the collector.
-               //if we don't, the collector may wait forever (==timeout) for non-existing acknowledges.
-               checkTopologyId(dwCommand);
-            }
-            if (trace) {
-               log.tracef("Command %s send to backup owner %s.", dwCommand.getCommandInvocationId(), backupOwners);
-            }
-            long sequenceNumber = triangleOrderManager.next(distributionInfo.getSegmentId(), dwCommand.getTopologyId());
-            BackupWriteRcpCommand backupWriteRcpCommand = commandsFactory.buildBackupWriteRcpCommand(dwCommand);
-            backupWriteRcpCommand.setSequence(sequenceNumber);
-            if (trace) {
-               logCommandSequence(id, distributionInfo.getSegmentId(), sequenceNumber);
-            }
-            // we must send the message only after the collector is registered in the map
-            rpcManager.sendToMany(backupOwners, backupWriteRcpCommand, DeliverOrder.NONE);
+         Collection<Address> backups = distributionInfo.backups();
+         if (backups.isEmpty()) {
+            return;
+         }
+         if (isSynchronous(dwCommand) || dwCommand.isReturnValueExpected()) {
+            commandAckCollector.create(id, rv, localAddress, backups, dwCommand.getTopologyId());
+            //check the topology after registering the collector.
+            //if we don't, the collector may wait forever (==timeout) for non-existing acknowledges.
+            checkTopologyId(dwCommand);
          }
+         sendToBackup(dwCommand, backups, distributionInfo.getSegmentId());
       });
    }
 
-   private void logCommandSequence(CommandInvocationId id, int segment, long sequence) {
-         log.tracef("Command %s got sequence %s for segment %s", id, sequence, segment);
+   private void sendToBackup(DataWriteCommand command, Collection<Address> backups, int segmentId) {
+      long sequenceNumber = triangleOrderManager.next(segmentId, command.getTopologyId());
+      BackupWriteRcpCommand backupWriteRcpCommand = commandsFactory.buildBackupWriteRcpCommand(command);
+      backupWriteRcpCommand.setSequence(sequenceNumber);
+      if (trace) {
+         logCommandSequence(command.getCommandInvocationId(), segmentId, sequenceNumber, backups);
```

Comment:
What does this pattern buy us? Doesn't the JIT eliminate everything inside `if (trace)` anyway?